### ISSUE TITLE
h2c support: upgrade insecure connections to http2

### ISSUE THREAD

**Issue by raboof** [akka/akka-http#1966]
There are 2 mechanisms for supporting http/2 on insecure connections: with prior knowledge (#1934) or with h2c using the HTTP Upgrade mechanism (http://httpwg.org/specs/rfc7540.html#discover-http).

We're unsure how widely used h2c is and are not currently planning to implement it, but if anyone would express interest or contribute it that would be great!

**Comment by satansk** [comment]
Let me try this :)

**Comment by dsilvasc** [comment]
It looks like h2c is the transport that the [grpc-java](https://github.com/grpc/grpc-java) client uses when connecting to a localhost server.

If you try to use Akka HTTP as a gRPC proxy, even with `akka.http.server.preview.enable-http = on` in your `application.conf`, Akka HTTP fails to handle gRPC requests with:

```
[my-system-akka.actor.default-dispatcher-9] [akka.actor.ActorSystemImpl(my-system)] Illegal request, responding with status '501 Not Implemented': Unsupported HTTP method: PRI
```

Documentation on the `PRI` method:
https://http2.github.io/http2-spec/#rfc.section.11.6

> This method is never used by an actual client. This method will appear to be used when an HTTP/1.1 server or intermediary attempts to parse an HTTP/2 connection preface.

**Comment by raboof** [comment]
It depends on what you set as https://grpc.io/grpc-java/javadoc/io/grpc/netty/NegotiationType.html : TLS is 'normal' HTTPS HTTP2, PLAINTEXT is #1934 and PLAINTEXT_UPGRADE is h2c.

@dsilvasc if you see that message in an Akka HTTP server it suggests the connection is 'raw' HTTP2 as supported by #1934, and not using the h2c upgrade mechanism.

**Comment by raboof** [comment]
@satansk did you have any luck with this so far?

**Comment by satansk** [comment]
@raboof not really, but I'm still working on this.

**Comment by raboof** [comment]
@satansk OK that's great! I'm also planning on having a look, so I'll share and we can compare notes.

**Comment by raboof** [comment]
I added some 'work in progress' in https://github.com/akka/akka-http/compare/h2c. I added a 'generic' upgrade mechanism next to the existing websocket upgrade mechanism. If this works out it would be interesting to see if we can remove the websocket-specific parts and have websockets support live on top of the generic support as well, in a binary compatible way. There's still some features and tests missing, but 'nghttp -v -u -H':method:GET' http://localhost:8080/whatever' already looks like it successfullly upgrades the connection (with some things missing, see the TODO's in that branch).
